File: nodes/hub/back/src/core/services/hub/automate.service.ts
import { BuildAgent, BuildConfig, DeployAgent, DeployConfig, Job } from "./types";
import { BuildAgentApi, DockerfilesConfigModel, DockerfilesConfigModelPlatformsEnum } from "../../apis/agent-build";
import { DeployJobModel, ProductionAgentApi } from "../../apis/agent-prod";
import { Log } from "../../utils/decorators/logger";
import { getLogger } from "../../utils/logger";
import { Service } from "@tsed/common";
import { AgentBuild } from "./agent/builder";
import { AgentDeployment } from "./agent/production";
import { QueueBuild } from "./queue/build";
import { JobBuild } from "./job/build";
import { QueueProduction } from "./queue/deployment";
import { JobDeployment } from "./job/deployment";

@Service()
export class AutomateService {
	private static log = getLogger.service(AutomateService);
	private services: {
		queues: { builds: QueueBuild; deployments: QueueProduction };
		jobs: { builds: JobBuild; deployments: JobDeployment };
		agents: { builds: AgentBuild; deployments: AgentDeployment };
	};

	constructor(agentBuild: AgentBuild, agentProduction: AgentDeployment, queueBuild: QueueBuild, queueproduction: QueueProduction, jobBuild: JobBuild, jobProduction: JobDeployment) {
		this.services = {
			agents: {
				builds: agentBuild,
				deployments: agentProduction,
			},
			queues: {
				builds: queueBuild,
				deployments: queueproduction,
			},
			jobs: {
				builds: jobBuild,
				deployments: jobProduction,
			},
		};
	}

	@Log(AutomateService.log)
	public async build(agent: BuildAgent, job: Job<BuildConfig>) {
		const platforms = Array<DockerfilesConfigModelPlatformsEnum>();
		for (const key in DockerfilesConfigModelPlatformsEnum) {
			if (job.config.dockerfiles?.platforms.some((p) => p === DockerfilesConfigModelPlatformsEnum[key])) {
				platforms.push(DockerfilesConfigModelPlatformsEnum[key]);
			}
		}

		job.startedAt = new Date();
		await this.services.jobs.builds.add(job);

		const dockerfiles: DockerfilesConfigModel | undefined = job.config.dockerfiles
			? {
					username: job.config.dockerfiles.username,
					files: job.config.dockerfiles.files.map((df) => ({
						image: df.image,
						path: df.path,
						tag: df.tag,
						wd: df.wd,
					})),
					platforms,
			  }
			: undefined;

		try {
			const { data: stdout } = await new BuildAgentApi(undefined, agent.uri).build({
				config: {
					...job.config,
					dockerfiles,
					bake: job.config.bake,
				},
				id: job.id,
			});
			job.stdout = stdout.join("");
		} catch (e) {
			const err = e as Error;
			const error = err.message + "\n" + err.stack;
			job.error = error;
			AutomateService.log.error(error);
		}

		job.finishedAt = new Date();
		this.services.agents.builds.finishJob(job.id);
		await this.services.jobs.builds.update(job.id, job);
	}

	@Log(AutomateService.log)
	public async deploy(agent: DeployAgent, job: Job<DeployConfig>) {
		job.startedAt = new Date();
		await this.services.jobs.deployments.add(job);
		if (job.config.docker != undefined) {
			try {
				const { data: stdout } = await new ProductionAgentApi(undefined, agent.uri).productionAgentBuild(job as DeployJobModel);
				job.stdout = stdout.join("");
			} catch (e) {
				const err = e as Error;
				const error = err.message + "\n" + err.stack;
				job.error = error;
				AutomateService.log.error(error);
			}
			this.services.agents.deployments.finishJob(job.id);
		}

		job.finishedAt = new Date();
		await this.services.jobs.deployments.update(job.id, job);
	}
}

File: nodes/hub/back/Core/Services/Hub/AutomateService.cs
using System.ComponentModel.DataAnnotations;
using Cicd.Hub.Abstractions.Common.Extensions;
using Cicd.Hub.Abstractions.Common.Helpers;
using Cicd.Hub.Abstractions.Interfaces.Repositories;
using Cicd.Hub.Abstractions.Interfaces.Services;
using Cicd.Hub.Abstractions.Interfaces.Watchers;
using Cicd.Hub.Abstractions.Transports.Agents;
using Cicd.Hub.Abstractions.Transports.Agents.Deploy;
using Cicd.Hub.Abstractions.Transports.Jobs.Build;
using Cicd.Hub.Abstractions.Transports.Jobs.Deploy;
using Cicd.Hub.Adapters.AgentBuildApi;
using Cicd.Hub.Adapters.AgentDeployApi;
using Cicd.Hub.Core.Assemblers;
using Microsoft.Extensions.Logging;
using AuthenticationApp = Cicd.Hub.Adapters.AgentDeployApi.AuthenticationApp;

namespace Cicd.Hub.Core.Services.Hub
{
	public class AutomateService : IAutomateService
	{
		private readonly IAuthenticationService authenticationService;
		private readonly BuildBakeAssembler buildBakeAssembler = new();
		private readonly DeployConfigAssembler deployConfigAssembler = new();
		private readonly GithubConfigAssembler githubConfigAssembler = new();
		private readonly JobAssembler.Build jobBuildAssembler = new();
		private readonly JobAssembler.Deploy jobDeployAssembler = new();
		private readonly IJobRepository jobRepository;
		private readonly IJobService jobService;
		private readonly ILogger<AutomateService> logger;

		public AutomateService(ILogger<AutomateService> logger, IJobRepository jobRepository, IAuthenticationService authenticationService, IDatabaseWatcher databaseWatcher, IJobService jobService)
		{
			this.logger = logger;
			this.jobRepository = jobRepository;
			this.authenticationService = authenticationService;
			this.jobService = jobService;
			databaseWatcher.WatchChanges();
		}


		public async Task<JobBuild> AskBuild(BuildConfig config, string userToken, Guid run)
		{
			logger.Enter($"{config.Github.Remote} {config.Github.Branch}");
			var appToken = await authenticationService.GetPermanentToken(userToken);
			var data = await jobService.Add(config, appToken, run);
			logger.Exit($"{config.Github.Remote} {config.Github.Branch}");
			return data;
		}

		public async Task<JobDeploy> AskDeploy(DeployConfig config, string userToken, Guid run)
		{
			logger.Enter($"{config.Url} {config.Docker.Compose.Path}");
			var appToken = await authenticationService.GetPermanentToken(userToken);
			var data = await jobService.Add(config, appToken, run);
			logger.Exit($"{config.Url} {config.Docker.Compose.Path}");
			return data;
		}

		public async Task Build(AgentBuild agent, JobBuild job)
		{
			logger.Enter($"{LogHelper.Get(agent.Id)} {LogHelper.Get(job.Id)}");

			job.StartedAt = DateTime.Now;
			await jobRepository.Update(job);

			var dockerfiles = job.Config.Dockerfile != null
				? new DockerfilesConfigModel
				{
					Platforms = job.Config.Dockerfile.Platforms,
					Files = job.Config.Dockerfile.Files.Select(df => new DockerFileConfigModel
							{
								Image = df.Image,
								Path = df.Path,
								Tag = df.Tag,
								Wd = df.WorkingDirectory
							}
						)
						.ToList(),
					Username = job.Config.Dockerfile.Username
				}
				: null;


			using var client = new HttpClient();
			var buildApi = new BuildAgentApi(agent.Url, client);
			var stds = await buildApi.BuildAsync(Adapters.AgentBuildApi.AuthenticationApp.CICD, new BuildConfigModel
				{
					Config = new()
					{
						Dockerfiles = dockerfiles,
						Github = githubConfigAssembler.Convert(job.Config.Github),
						Bake = job.Config.Bake != null ? buildBakeAssembler.Convert(job.Config.Bake) : null
					},
					Id = job.Id.ToString()
				}, job.Token
			);
			job.FinishedAt = DateTime.Now;

			var stdout = "";
			var stderr = "";
			stds.ForEach((std, i) => {
					var currentDockerFile = dockerfiles?.Files.ToList()[i];
					var header = "";
					if (currentDockerFile != null)
					{
						header = $"### {currentDockerFile.Image}:{currentDockerFile.Tag ?? "latest"} at {currentDockerFile.Wd}{currentDockerFile.Path} ###\n";
					}

					if (std.Status == 0)
					{
						stdout += header;
						stdout += std.Stderr;
					}
					else
					{
						stderr += header;
						stderr += std.Stderr;
					}
				}
			);
			job.Stderr = stderr;
			job.Stdout = stdout;


			await jobRepository.Update(job);

			jobService.SetJobCompleted(job.Id);

			logger.Exit($"{LogHelper.Get(agent.Id)} {LogHelper.Get(job.Id)}");
		}

		public async Task Deploy(AgentDeploy agent, JobDeploy job)
		{
			logger.Enter($"{LogHelper.Get(agent.Id)} {LogHelper.Get(job.Id)}");
			job.StartedAt = DateTime.Now;
			await jobRepository.Update(job);
			using var client = new HttpClient();
			var api = new DeployAgentApi(agent.Url, client);

			var stds = await api.DeployAsync(AuthenticationApp.CICD, new DeployJobModel
				{
					Config = deployConfigAssembler.Convert(job.Config),
					Id = job.Id.ToString()
				}, job.Token
			);
			job.FinishedAt = DateTime.Now;

			var stdout = "";
			var stderr = "";
			stds.ForEach((std, i) => {
					var currentDockerFile = job.Config.Docker.Compose.Path;
					var header = $"### {currentDockerFile} ###\n";
					if (std.Status == 0)
					{
						stdout += header;
						stdout += std.Stderr;
					}
					else
					{
						stderr += header;
						stderr += std.Stderr;
					}
				}
			);
			job.Stderr = stderr;
			job.Stdout = stdout;
			await jobRepository.Update(job);

			jobService.SetJobCompleted(job.Id);

			logger.Exit($"{LogHelper.Get(agent.Id)} {LogHelper.Get(job.Id)}");
		}
	}
}